projects/libevent/bufferevent_fuzzer.cc

```diff
@@ -0,0 +1,116 @@
+/* Copyright 2023 Google LLC
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string>
+#include <sys/socket.h>
+
+#include <fuzzer/FuzzedDataProvider.h>
+
+extern "C" {
+#include "libevent/include/event2/buffer.h"
+#include "libevent/include/event2/bufferevent.h"
+#include "libevent/include/event2/event.h"
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+
+  FuzzedDataProvider data_provider(data, size);
+
+  std::string s1 = data_provider.ConsumeRandomLengthString();
+  std::string s2 = data_provider.ConsumeRandomLengthString();
+  size_t int1 = data_provider.ConsumeIntegral<size_t>();
+  size_t int2 = data_provider.ConsumeIntegral<size_t>();
+  size_t int3 = data_provider.ConsumeIntegral<size_t>();
+  size_t int4 = data_provider.ConsumeIntegral<size_t>();
+
+  int use_pair = int1 % 2;
+  int options1 = int2 % 16;
+  int options2 = int3 % 16;
+
+  struct bufferevent *bev1 = NULL, *bev2 = NULL, *pair[2];
+  struct event_base *base = NULL;
+  struct evbuffer *evbuf = NULL;
+  static struct ev_token_bucket_cfg *conn_bucket_cfg = NULL;
+  struct bufferevent_rate_limit_group *bev_rate_group = NULL;
+  char buf[128];
+
+  /*create a buffer event*/
+  base = event_base_new();
+  if (use_pair == 0) {
+    if (bufferevent_pair_new(base, options1, pair) == -1) {
+      event_base_free(base);
+      return 0;
+    }
+    bev1 = pair[0];
+    bev2 = pair[1];
+    bufferevent_pair_get_partner(bev1);
+  } else {
+    bev1 = bufferevent_socket_new(base, -1, options1);
+    bev2 = bufferevent_socket_new(base, -1, options2);
+  }
+
+  /*bufferevent_filter_new*/
+  bev1 = bufferevent_filter_new(bev1, NULL, NULL, options1, NULL, NULL);
+  bev2 = bufferevent_filter_new(bev2, NULL, NULL, options2, NULL, NULL);
```

Comment:
I don't think the leaks from the CI are valid leaks. I think they happen because you re-assign `bev2` here (previously assigned on line 57) and that overwrites a pointer to some dynamic memory. Thus, a point is left hanging -- I think the fuzzer is missing some cleanup?

Reply:
@DavidKorczynski I really appreciate your help. Indeed, it was an issue with hanging pointers on lines 65 and 66. I fixed the fuzz target and now it works fine, as well I resolved the issue with double free.

I did not use your fix, because **bufferevent_filter_new** function accepts a non-NULL **bufferevent** object, but it helped me to realize where the problem was. In details, I've added two **bufferevent** objects to be assigned from **bufferevent_filter_new** function and set **conn_bucket_cfg** to NULL to fix potential double free. Please approve when you have time. Thanks!